Feature/check urls validity

davidlonjon opened the pull request:
This PR works with the https://github.com/Automattic/jetpack-boost/pull/960 PR. If this is accepted, this PR will need to be merged first and then the PR above will have to be updated to revert back changes made to the `package.json` and `package-lock.json` files.

This PR does allow for the `generateCriticalCSS` function to accept an additional successTargets parameter which will allow to make sure it can identify and work with a subset of urls from the ones provided to generate Critical CSS.

This `successTargets` data is provided by the plugin and contains 2 properties:
- `max`: The max number of urls that could be returned by the provider.
- `min`: The required number of valid urls for the provider in order to successfully generate Critical CSS for the provider.

Note that if the `successTargets` data is not provided, the `generateCriticalCSS` function will try to generate Critical CSS for all the urls provided.

However if these 2 properties are provided, it will try to identify which urls are the invalid or valid ones in order to get enough valid urls to generate Critical CSS for that set of urls. The number of valid urls required is controlled by `min`.

It is also important to note that there could be some cases where the number of urls provided is less than the `max` and `min`. In that case, `min` is being re-calibrated.
For example a site may have only than 6 posts but `max` is 20 and `min` was 10. In that example the number of required valid urls will become 3 (6 / (20/10)). This logic could be re-adjusted.

davidlonjon committed:
Fix typo

davidlonjon committed:
Add logic to check for urls validity using a threshold among a pool of urls

davidlonjon committed:
Simplify logic to check for valid urls

davidlonjon committed:
Remove console statement

davidlonjon committed:
Rename meta properties name and variable to clarify meaning

davidlonjon committed:
Add missing logical check

davidlonjon committed:
Fix logic to re-calibrate the original number of urls.

thingalon commented on lib/generate-critical-css.js:
```diff
@@ -8,12 +10,15 @@ async function generateCriticalCSS( {
 	urls,
 	viewports,
 	filters,
+	meta,
```
Instead of `meta.required_valid_urls_number` and `meta.max_urls_number`, I would prefer more contextually meaningful variable names.

How about:
```
successTargets: { min: 5, max: 20 },
```

i.e.: I want at least 5 successes, but stop after 20 successes

What do you think?

davidlonjon replied:
@thingalon, thanks, I was struggling the naming, these makes sense. I did some renaming in this repo in commit c5a9fac81a55e7b3d47f506b78faea4e9163bbcd and in https://github.com/Automattic/jetpack-boost/pull/960/commits/7841160a46275d32c29b6341d42184b42a718714

thingalon commented on lib/generate-critical-css.js:
```diff
@@ -8,12 +10,15 @@ async function generateCriticalCSS( {
 
 	try {
+
+		urls = await getValidUrls(urls, meta, browserInterface);
```
🤔  Now I see this implementation, I don't think we need a 3rd URL loading pass upfront to check for URL validity.

In this PR, you're passing in a minimum and maximum number of successful page-loads / generation attempts. It should be easy to loop through all provided URLs in the [main loop](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/trunk/lib/generate-critical-css.js#L41) and abort early once enough urls have been successfully processed. At the end of the loop, it can check if enough urls were successfully processed and throw an error if not.

davidlonjon replied:
@thingalon ,

I have been struggling with it which is that the `loadPage` [method ](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/browser-interface-iframe.js#L120) is using throw statement to stop the whole process early if an url in the pool of urls has an error. 

Since `loadPage` is effectively called 3 times in the `generateCriticalCSS` [function](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L7) via 

A: `await CSSFileSet.collate( browserInterface, urls );` ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L26))
B: `const pageSelectors = await browserInterface.runInPage(` ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L48))
C: const pageAboveFold = await browserInterface.runInPage(' ([link](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/feature/check-urls-validity/lib/generate-critical-css.js#L76))

There are actually 2 loops too, the one in **A**, and the one that encompass **B** and **C**.

Regardless of the 2 loops which was one hurdle for me, the `throw` statements from the `loadPage` which is bubbled up to the `generateCriticalCSS` was the second hurle. If we have 10 urls and want 5 successful one, if 1,2,3,4 urls are failing, we still want to carry on the process for url 5, 6, 7, 8 and 9.

So I decided to use `getValidUrls` to go throw the whole pool of urls and identify the valid and invalid one without raising an error and just generating the valid set of urls which can then be used in the next 2 loops.

Not sure if it makes sense, I might still not be seeing the easier path for implementation here without a bigger refactoring. I think at this stage it would be useful for me to have a video chat and session sharing with you.

thingalon reviewed (changes requested):
Now that I see this implementation (and the parameters you have proposed to pass to the library), I recommend a few changes. I don't think we need to attempt to load all the URLs upfront to find valid ones, and I think the min/max variables could be better named.

davidlonjon committed:
Rename parameters and object properties to clarify meaning

pyronaur commented on lib/browser-interface-iframe.js:
```diff
@@ -63,6 +63,22 @@ class BrowserInterfaceIframe extends BrowserInterface {
 	}
 
+	async checkUrlValidity( pageUrl ) {
+		const checkResult = {
```
I read `check-result` as "perform a check on the result" in code. Perhaps we can just simplify and call it `const result`?

pyronaur commented on lib/generate-critical-css.js:
```diff
@@ -1,3 +1,5 @@
+const {getValidUrls} = require( './helpers' );
```
Missing spaces around `{` `}`

pyronaur commented on lib/helpers.js:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @returns {string[]}
+ */
+async function getValidUrls (urls, successTargets, browserInterface) {
```
Looks like we're filtering URLs here. Perhaps `filterValidUrls` instead?

pyronaur commented on lib/helpers.js:
```diff
@@ -0,0 +1,51 @@
+
+	if (failures.length > 0 && minSuccessTarget > validUrls.length) {
+		// If we have not enough valid urls from the the pool of urls, throw an error
+		// TODO: Maybe create a new error Class which can gather the multiple failures.
```
TODOs should be GitHub issues.

pyronaur reviewed (changes requested):
Left a few code style related comments

Commit:
Detect URL errors while collating CSS URLs

davidlonjon committed:
Merge branch 'trunk' into feature/check-urls-validity

# By pyronaur (5) and others
# Via David Lonjon (2) and GitHub (2)
* trunk:
  Fixed the unit test
  Add prettier as a dependency
  More ESLint fixes
  Fix ESLint Issues
  Add WordPress ESLint stnadards
  Update packages
  Add editor config
  add getType to index
  Ensure that errors retain their full name in bundled / minified versions
  Add tool to bundle library into standalone loadable js lib

# Conflicts:
#	lib/css-file-set.js

davidlonjon committed:
Fix coding standards

davidlonjon committed:
Fix issue with setting object property

davidlonjon committed:
Fix issue with the SuccessTargetError constructor

davidlonjon committed:
Fix gathering error messages

davidlonjon committed:
Refactor the logic to check for having enough valid urls

davidlonjon committed:
Add a method to get the error data

davidlonjon commented:
@thingalon,

Thank you so much for the refactoring. It is much more efficient and elegant 👍 🙇 

I have fixed a few things and re-added the logic to support the case where the total number of urls from the pool is less than the min threshold. in that case i am re-calibrating the min threshold.

I have now also updated the https://github.com/Automattic/jetpack-boost/pull/960 PR to support the new `SuccessTargetError` error and to allow displaying correctly its child page errors such as 

![image](https://user-images.githubusercontent.com/927932/119824983-727c8d00-bf31-11eb-9759-2d5c393d3a12.png)

The UI might need to be reviewed. I just left it very basic right now in purpose.

@pyronaur, I believe that Mark has addressed all your original feedback.

Commit:
Keep the logic to ensure a sensible minimum successTarget simple'

Commit:
Enforce max urls if provided

Commit:
Add some tests to verify the minimum success targets are properly handled and respected

Commit:
Add a new test which demonstrates max is not 100% respected

Commit:
Do not load more pages that needed to hit the success max

thingalon commented:
Hi @davidlonjon, I made a few improvements to your PR:

1. I've simplified the logic to ensure that `successTargets.min` is not lower than the initial URL count. It's now just a simple condition at the start of `generateCriticalCSS`; I don't think that we should make complex guesses as to the behaviour the caller wants in this library.
2. I've added code to enforce the `successTargets.max` if specified, and
3. I've added tests to verify that `successTargets.min` and `successTargets.max` behave as expected.

Commit:
Clean up minTarget handling; don't modify params

davidlonjon committed:
Merge branch 'trunk' into feature/check-urls-validity

* trunk:
  add an empty npmignore
  Add a main
  Add prepare script to build bundle

davidlonjon committed:
Refactor to replace success targets by a ratio

davidlonjon committed:
Add missing comma

davidlonjon commented on lib/generate-critical-css.js:
```diff
@@ -1,26 +1,84 @@
-	if ( ! progressCallback ) {
-		progressCallback = () => {};
-	}
+	// TODO: Review if we want to be less strict and use Math.floor instead.
```
@thingalon , @pyronaur : I am not sure if we want to be more conservative using `ceil` which means that the success urls threshold is one url more, or a little more tolerant and use `floor`

thingalon replied:
We certainly don't want to use `Math.floor` here. If there is 1 URL in the set, then it will always set the threshold at 0 ;)

I think that erring on the side of conservative is fine. If any callers are more fussy than that, they can always bump their success threshold down.

davidlonjon replied:
Sounds good. Let's stick with `ceil` :)

thingalon commented on lib/generate-critical-css.js:
```diff
@@ -1,26 +1,84 @@
+
+			await cssFiles.addMultiple( url, absoluteIncludes );
+
+			// Abort early if we hit the threshold of success urls.
```
Now that we're using a success threshold instead of a min/max, I no longer think that `collateCssFiles` should abort early once the minimum success threshold has been met. There are some critical problems that might arise here:

1. If the first URL is successfully read, then the success rate will be 100% at this stage, and it will happily abort early. Then, no other URLs will be examined.
2. If a URL works at this stage, but fails at a later stage, it will have been removed from the valid URL set - and the overall process will fail, even if another URL in the set might be valid.

IMHO, now that we're using success thresholds we should simplify and count successes in the main processing loop - then evaluate if the threshold has been met at the end.

We can't abort early just based on a success threshold %, unless we _also_ have a minimum absolute number of URLs that must be successfully processed too. Otherwise we'll stop after the first sucess, and that would be bad. :)

thingalon replied:
Oh, I retract the above statements. I realized after posting this that you are calculating an absolute number of urls that must succeed based on the number supplied, and so my comments about bailing after one are off the mark!

thingalon reviewed (changes requested):
See inline comment.

We can't afford to abort early based on a success % threshold.

thingalon reviewed (approved):
On reflection, my previous review was based on a mis-reading of what's here. I think this works, and can go in.

davidlonjon commented:
> On reflection, my previous review was based on a mis-reading of what's here. I think this works, and can go in.

🙏🏻

davidlonjon committed:
Remove todo comment

davidlonjon committed:
Add a watch command

davidlonjon commented:
@thingalon,

I have removed the TODO comments and also added a `watch` command (dc0c0e6) to make development with the plugin ( i have a bit of a hack from the plugin side to make it work) easier.

Commit:
track child errors explicitly, guarantee their type

Commit:
Linty McLintface

davidlonjon committed:
Merge branch 'trunk' into feature/check-urls-validity

* trunk:
  Lint fixes
  Absolute-ify URLs before generating Critical CSS
